File: web/src/utils/ttySanitize.ts
/** Strip PTY/CLI control noise (ANSI cursor toggles, bare CSI) from auth/terminal text. */

/* eslint-disable no-control-regex -- intentional ANSI / control char scrubbing */
const ANSI_OR_CSI = /\x1b(?:[@-Z\\-_]|\[[0-9;?]*[ -/]*[@-~])/g;
/* eslint-enable no-control-regex */

const BARE_BRACKET_SEQ = /\[[0-9;?]*[a-zA-Z]/g;

export function stripTerminalControlSequences(text: string): string {
  return text
    .replace(ANSI_OR_CSI, "")
    .replace(BARE_BRACKET_SEQ, "")
    .replace(/\r/g, "");
}

File: src/agent_lab/tty_sanitize.py
"""Strip PTY/CLI control noise from text shown in the UI."""

from __future__ import annotations

import re

# ESC … final byte (CSI, OSC, etc.)
_ANSI_ESCAPE = re.compile(r"\x1b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])")
# Bare CSI when ESC was dropped (e.g. "[?25h" show-cursor from Claude/Codex CLI)
_BARE_CSI = re.compile(r"\[[\?0-9;]*[a-zA-Z]")


def sanitize_tty_text(text: str) -> str:
    """Remove ANSI cursor/screen control sequences from PTY output."""
    if not text:
        return text
    cleaned = _ANSI_ESCAPE.sub("", text)
    cleaned = _BARE_CSI.sub("", cleaned)
    return cleaned.replace("\r", "")
